datapath: Support NodePort BPF on L2-less devices

brb opened the pull request:
See commit msgs.

I haven't added IPv6 tests, as I'd need to extend https://github.com/cilium/kube-wireguarder to support it. At this point, it would be a waste of time, as we are planning to add the native Wireguard support in this release cycle. So we can add the tests once https://github.com/cilium/cilium/issues/15075 has been resolved.

To fix L2-less with the fast redirect I've opened https://github.com/cilium/cilium/issues/15075.

Fix #12317

```release-note
Add NodePort BPF support to L2-less devices (wireguard, tun, etc)
```

brb committed:
datapath: Support NodePort BPF on L2-less devices

This commit extends NodePort BPF by making it possible to run it on
L3 network devices (without L2 addr). One prominent case is the
Wireguard tunnel device (wg0).

The main idea of the change is to make ETH_HLEN configurable via ELF
templating (on L2-less devices we set it to 0 during the load time),
and to craft an L2 hdr when forwarding from L2-less to L2 device.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Do not ignore L2-less devices in dev auto detection

The previous commit added a support for L2-less devices.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Fix ordering of BPF host routing detection

Fix a bug when NodePort BPF is disabled after the device
detection, and BPF host routing is kept enabled.

Fixes: 7e0cb333 ("bpf: do not enable host routing when kpr is disabled")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm reviewed (changes requested):
Changes look good overall.

Needs documentation and disclaimers that it will likely won't work for large clusters (>500 or >1000 nodes) given that kube-wireguarder is watching for all events in CiliumNodes and K8s Nodes.

brb committed:
datapath: Only make ETH_HLEN as static data if L3 dev exist

This should eliminate a datapath perf penalty introduced by the L2-less
changes when running on systems which all involved devices have L2
addrs.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann commented on bpf/lib/common.h:
```diff
@@ -118,6 +119,17 @@ static __always_inline bool validate_ethertype(struct __ctx_buff *ctx,
 	struct ethhdr *eth = data;
 
+#if __ctx_is == __ctx_skb
+	/* XDP does not support packets without L2 hdr, so skip the check.*/
```
Given it's not possible to attach to XDP for those devices, why do we need the extra `__ctx_is == __ctx_skb` check?

brb replied:
I wanted to compile out the case for XDP to save a few CPU cycles. Let me check whether compiler does it for us.

brb replied:
OK, just checked - this one gets compiled away :+1:

borkmann commented on bpf/lib/common.h:
```diff
@@ -146,10 +160,18 @@ __revalidate_data_pull(struct __ctx_buff *ctx, void **data_, void **data_end_,
+__revalidate_data_pull(struct __ctx_buff *ctx, void **data, void **data_end,
+		       void **l3, const __u32 l3_len, const bool pull)
+{
+	return __revalidate_data_pull_with_eth_hlen(ctx, data, data_end, l3,
```
small nit: maybe just `____revalidate_data_pull`

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -161,6 +161,37 @@ static __always_inline bool dsr_is_too_big(struct __ctx_buff *ctx __maybe_unused
+		 __u32 ifindex __maybe_unused,
+		 bool *l2_hdr_required __maybe_unused)
+{
+/* XDP does not support L2-less devices, so skip the checks. */
```
For XDP, shouldn't these just be compiled out w/o the `__ctx_is == __ctx_skb` test? (`IS_L3_DEV` always false, `ETH_HLEN` always 14)

brb replied:
Let me check whether compiler removes them.

brb replied:
OK, this one gets compiled out too. I'm going to drop the redundant `__ctx_is` checks.

borkmann replied:
Sounds good, the less such checks where not needed the better.

borkmann commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -448,6 +415,43 @@ func finishKubeProxyReplacementInit(isKubeProxyReplacementStrict bool) {
 
 	// After this point, BPF NodePort should not be disabled
 
```
I think we need this as separate fix for 1.9 stable tree as well? So this should probably be broken out and then cherry-picked as single commit to 1.9.

brb replied:
Yeah, I have it in my TODO.

borkmann commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -448,6 +415,43 @@ func finishKubeProxyReplacementInit(isKubeProxyReplacementStrict bool) {
+			!option.Config.EnableBPFMasquerade:
+			msg = fmt.Sprintf("BPF host routing requires %s.", option.EnableBPFMasquerade)
+		case !hasHardwareAddress():
+			msg = fmt.Sprintf("BPF host routing is incompatible with devices without L2 addr.")
```
Could you elaborate on why incompatible? I think it should be working just that we need to push a L2 header before the redirect_peer? Although I'm not even sure if the stack cares about that even. Maybe needs the same change as in https://github.com/cilium/cilium/pull/15148 to get it working?

brb replied:
> I think it should be working just that we need to push a L2 header before the redirect_peer?

As far as I've checked the code path, it should be enough. I didn't want to add to this PR to avoid making it too complex.

borkmann replied:
Ok, if it's follow-up, lets just say `currently not supported` rather than `incompatible`. Hopefully for 1.10 scope, we'll resolve it in time. I can also help if needed.

borkmann commented on bpf/include/bpf/helpers_skb.h:
```diff
@@ -60,4 +60,7 @@ static int BPF_FUNC(sk_release, struct bpf_sock *sk);
 		    __u64 flags);
 
+static int BPF_FUNC(skb_change_head, struct __sk_buff *skb, __u32 head_room,
+		    __u64 flags);
```
nit: should go below `/* Packet manipulation */`

borkmann commented on bpf/include/linux/if_ether.h:
```diff
@@ -29,7 +29,7 @@
 
 #define ETH_ALEN	6		/* Octets in one ethernet addr	 */
-#define ETH_HLEN	14		/* Total octets in header.	 */
+#define __ETH_HLEN	14		/* Total octets in header.	 */
```
Hm, I missed this earlier, but we should not mangle headers that we sync from upstream kernel. (bpf/include/linux/*)

brb replied:
Ouch, then we would need to rename `ETH_HLEN` in our datapath code :-/

borkmann replied:
Hm :/ yeah I guess it would need a `L2_HLEN`...

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -653,6 +692,16 @@ int tail_nodeport_nat_ipv6(struct __ctx_buff *ctx)
+	if (ret != 0)
+		goto drop_err;
+	if (!l2_hdr_required)
+		goto out_send;
```
nit: the __maybe_unused from `out_send: __maybe_unused` can then probably be removed.

borkmann reviewed (commented):
modulo the minor bits, lgtm!

christarazi commented on pkg/datapath/linux/config/config.go:
```diff
@@ -510,6 +517,30 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	return fw.Flush()
 }
 
+func isL3DevMacro() (string, error) {
```
Might be worth briefly documenting this and why it exists

christarazi commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -449,6 +415,45 @@ func finishKubeProxyReplacementInit(isKubeProxyReplacementStrict bool) {
 
 	// After this point, BPF NodePort should not be disabled
 
+	if !option.Config.EnableHostLegacyRouting {
```
Maybe it makes sense to make a separate commit to move this entire block, then a new commit adding the change you want, so that we don't lose the history / break git bisect?

christarazi commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -434,7 +435,7 @@ func finishKubeProxyReplacementInit(isKubeProxyReplacementStrict bool) {
 			!option.Config.EnableBPFMasquerade:
 			msg = fmt.Sprintf("BPF host routing requires %s.", option.EnableBPFMasquerade)
-		case !hasHardwareAddress():
+		case !mac.HaveMACAddr(option.Config.Devices):
```
If you break apart the commits like mentioned previously, then this commit can be squashed into the second commit, and this delta between commits won't exist.

christarazi reviewed (commented):
:tada: Awesome! Just a few comments below

brb committed:
daemon: Disable BPF host routing when L2-less detected

Until https://github.com/cilium/cilium/issues/15075 has been resolved.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -161,6 +161,32 @@ static __always_inline bool dsr_is_too_big(struct __ctx_buff *ctx __maybe_unused
+
+		if (eth_store_proto(ctx, proto, 0) < 0)
+			return DROP_WRITE_ERROR;
+	}
```
AFAICS, the `ETH_LEN==0` case will maximize complexity here, so we might want to set that macro to 0 in `bpf/Makefile`.

brb replied:
https://github.com/cilium/cilium/pull/15343

pchaigno commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -449,6 +415,43 @@ func finishKubeProxyReplacementInit(isKubeProxyReplacementStrict bool) {
+		msg := ""
+		switch {
+		case !option.Config.EnableNodePort:
+			msg = fmt.Sprintf("BPF host routing requires %s.", option.EnableNodePort)
```
AFAICS, we'll never hit this case because, as said above:
> After this point, BPF NodePort should not be disabled

So if NodePort is disabled for some reason other than explicitly disabled, `ENABLE_REDIRECT_FAST` will remain enabled in the datapath.

brb replied:
This has resolved the issue https://github.com/cilium/cilium/pull/15240/files.

pchaigno reviewed (commented):
Couple comments which we'll require follow ups if I didn't miss something.
